### ISSUE TITLE
After TB68 update: multiple accounts from same server no longer working

### ISSUE THREAD

**Issue by stoppeye** [jobisoft/DAV-4-TbSync#127]
After upgrading to TB68 (and TbSync 2.4 + CalDav Provider 1.3), I can't get my setup with two CalDav calendars from same server (same location, but two different accounts on a SmarterMail mail server) to work.

Previously, I used the setup as described in the Wiki pages and it worked perfectly.

After the update, I tried a fresh start (unsubscribed all calendars in lightning + deleted all accounts in TbSync). When I add the first calendar (business), all is fine. I see the added calendar in lightning with the name I provided in the accout setup. But as soon as I add the second calendar/account (private), the first one gets replaced in lighting with the second one. I suddenly get business events in my private calendar (according to the labeling in Lightning...).

As a workaround, I now use two different domain names for the originating server. But since only the main domain on the server has an SSL certificate, I can only set up one of my calendars with HTTPS.

And the passwords don't get saved. Every morning I have to enter my calendar passwords. And obviously I check the option to save the passwords in Password Manager every time. But if I check the saved passwords in the TB security settings, I can see that they don't show up.

And related to this problem: on the [Wiki page](https://github.com/jobisoft/DAV-4-TbSync/wiki/Adding-multiple-accounts-from-the-same-server) it says "This has been fixed in Thunderbird 68!". What has been fixed? Are those advanced config options no longer required? Has the problem with multiple accounts from same server been fixed? According to my experience, it rather seems that things are broken in TB68, not fixed...

**Comment by jobisoft** [comment]
Yes, this has been fixed in TB68 (by myself) and the workarounds are no longer needed. I am astonished, that it does not work for you. If you edit the calendar settings in lightning you will see this:

![image](https://user-images.githubusercontent.com/5830621/65675328-474fe500-e04e-11e9-9876-cb0e82fa1753.png)

And there is now a username field and if TbSync sets these calendars up, he adds the username there. Are they there? Different for your two calendars?

**Comment by jobisoft** [comment]
It all works as expected here (ownCloud server):

![image](https://user-images.githubusercontent.com/5830621/65677384-c692e800-e051-11e9-958e-f5a9840ad50c.png)

![image](https://user-images.githubusercontent.com/5830621/65677404-d27eaa00-e051-11e9-96a0-8c1246865cde.png)

![image](https://user-images.githubusercontent.com/5830621/65677440-e0ccc600-e051-11e9-9d83-d526d6e302ed.png)

I am running a slightly updated (beta) version of TbSync here, but it should not make a difference.

**Comment by stoppeye** [comment]
Lightning won't let me add a new calendar, if the location is the same as in another calendar (even if the username is different):

![screenshot--2019 09 26-11_34_33](https://user-images.githubusercontent.com/7151694/65677503-00fc8500-e052-11e9-88a5-42aed4766bd3.png)

**Comment by jobisoft** [comment]
Well, you cannot add the same (!) calendar with different username. 

Each user has its own calendar, right? On the same server, but both users still have independent calendars with different Urls, right?

**Comment by stoppeye** [comment]
That's my whole problem: SmarterMail won't give my individual calendar URLs. There is a function inside the SmarterMail BE where you can look up the CalDav URLs for a specific account. If there is only one calendar, it's automatically the primary calendar. So the information I get looks like this:

```
All Calendars (Autodiscover):
https://hostname:443/WebDAV/

My Calendar (Primary):
https://hostname:443/WebDAV/cal/default/
```

And if I go to a different account on the same server, it will look exactly the same. I know it's stupid, but I haven't found a way to get around this. (In older versions of SmarterMail you could find the UUID of a calendar in the respective XML file in you account and construct your own calendar URL from that, but that doesn't seem to work anymore.)
That's why I have to work around this by using a different hostname, so that I will have a different URL.

That's the reason I was so happy with TbSync, because pre TB68 I could finally get this to work.

**Comment by jobisoft** [comment]
Ok, remove the calendar you set up manually and let TbSync set it up for one of your accounts. What URL did TbSync add? You can see it in the calender properties dialog as shown in one of my very first screenshoot?

**Comment by jobisoft** [comment]
Explenation: TbSync does resource discovery. So the URL you enter in TbSync account setup is not the url TbSync will use to add your calendars, but it will actually lookup the individual calendar URLs and in the settings dialog of lightning, you can see the URL Tbsync has found.

I will close this if no further feedback.

**Comment by stoppeye** [comment]
After using the automatic setup in TbSync, the URL added to the calendar (in Lightning) is:

https://hostname/WebDAV/cal/default/

That's the same for both accounts. (The only difference being, that I use different hostnames.)

PS: sorry for the late feedback, but I have only upgraded my office PC to TB68 so far + I wasn't at the office yesterday afternoon

**Comment by jobisoft** [comment]
I think that is wrong as CalDAV Urls must be unique, I think.

@dilyanpalauzov: Do you know any details here? The user is getting the same resource URLs for all users and the server is only looking at the authenticated username to pick the calendar. Is that allowed? That breaks lightning core. Is there an RFC about that?

**Comment by stoppeye** [comment]
BTW: I just noticed that you can "cheat" lightning into accepting the "same" location URL for different calendards by just adding a random parameter at the end (eg. ?1234). At least that allows me to run all my calendards with https.

**Comment by jobisoft** [comment]
I will add that as a hidden property, so you can continue to use TbSync.

**Comment by stoppeye** [comment]
That still leaves me with the password problem. Looking at my saved passwords, I have noticed two things:

![screenshot--2019 09 27-09_46_02](https://user-images.githubusercontent.com/7151694/65752193-8c3a5100-e10c-11e9-820d-66d687c41128.png)

1. The passwords stored by TbSync/Dav used to have the protocol prefix dav: in previous versions. This seems missing here.
2. The URL stored with the password is the same as shown in the account settings in TbSync, but not the same as stored in the calendar location field in Lightning

I suppose that's the reason why the stored calendar passwords aren't found and why I have to enter them every time I start up Thunderbird. I just wonder why the passwords aren't stored after that, even though I click the checkbox to have them stored by Password Manager?

Should this go into a new issue?

**Comment by jobisoft** [comment]
The password lookup is using the username and it should work. Can I have a test account on that server? That would speed things up.

**Comment by stoppeye** [comment]
Ok, I'll send you the login credentials by email.

**Comment by dilyanpalauzov** [comment]
The question to me is, whether the same Calendar-URL can be returned to different users.

As a rule of thumb, the accounts must co-exist independently of each other.  If a CUA has several accounts, these can be for the same user or not, the server can return any data, that is valid within that account.

A user can make a personal calendar shared.  Then the same calendar is visible from many accounts and if the server returns for all accounts the same URL for the same calendar, the CUA can recognize this and synchronze the data only once.

There are two standard approaches to make calendar shared
* By granting access to the webdav collection of the shared calendar and the calendar-home-set of the user, who wants to share the calendar.  Then the CUA looks with the DAV:princapal-property-search report for the other user and subscribes to its calendar
* With webdav notifications, described at https://evertpot.com/webdav-caldav-carddav-sharing/

Since no open-source CUA implements either of these methods, the providers have to find a way how to provide the users with shared calendars, despite of this.  The way is to return many calendar-home-sets, when a user logs in and then to return in each home-set the shared calendars.  At the end, the CUA handles then all calendars as equivalent: it does not know, which calendar belongs to the use, and which is shared.

**Comment by jobisoft** [comment]
I impl the first method, don't I? (but OT). If the server has shared resources, I show them.

But I read you correctly, that this is indeed valid?

**Comment by dilyanpalauzov** [comment]
The first method is not implemented in DAV-4-TbSync.  DAV-4-TbSync does not perform the DAV:principal-property-search report.  In the first approach, the CUA must remember to which calendars a user has subscribed and these are then queried every time.  The task of the server is only to offer the client to possibility to discover calendars of other users, not to store to which calendar a user has subsribed.

The webdav-Notification method is supposed to supersede the DAV:principals-properties-search REPORT method concerning shared collections.

**Comment by jobisoft** [comment]
Password fails because your server returns an empty realm:

WWW-Authenticate | Basic realm=""

Oh boy, why? Lightning does not allow to store password for empty realms, so you will be prompted on every restart.

**Comment by jobisoft** [comment]
I updated the wiki page:
https://github.com/jobisoft/DAV-4-TbSync/wiki/Adding-multiple-accounts-from-the-same-server

Liek to give it a try? Before upgrading to beta, disable all your tbsync accounts. Itis not suggested to downgrade fro beta back to stable. So you better try it out in a fresh profile.

**Comment by stoppeye** [comment]
Oh boy, you're too fast for me... ;-)

I will try this, but I can't do it today. I'll come back to you.

Regarding the empty basic realm: Thanks for this information! I'll check whether this can be fixed on the server.

**Comment by stoppeye** [comment]
I finally installed today's betas. I was able to create a calendar with an identical URL to an existing calendar (with a different username) and it seems to work fine. The funny thing is: this worked even BEFORE I added extensions.dav4tbsync.enforceUniqueCalendarUrls = true in the advanced config.

**Comment by stoppeye** [comment]
Regarding Lightning not storing passwords for an empty realm: has this changed with TB68? Because I didn't have issues with that before the upgrade and the mail server certainly didn't change in that regard.

**Comment by jobisoft** [comment]
I just checked with TB60. And you are right. With TB60 lightning replaced empty realms with the url to be able to store the password. It no longer does that.

**Comment by jobisoft** [comment]
Regarding `extensions.dav4tbsync.enforceUniqueCalendarUrls`. Sure you did not manipulate the other url by hand? Could you remove both calendars/accounts and set them up again with that option disabled? Does it really work without the tweak?
